src/main/java/com/linecorp/armeria/common/stream/DefaultStreamMessage.java

```diff
@@ -309,24 +309,26 @@ private boolean notifyCompletableFuture(Queue<Object> queue) {
     }
 
     private void notifySubscriberWithCloseEvent(Subscriber<Object> subscriber, CloseEvent o) {
-        setState(State.CLEANUP);
-        cleanup();
-
-        final Throwable cause = o.cause();
-        if (cause == null) {
-            try {
-                subscriber.onComplete();
-            } finally {
-                closeFuture.complete(null);
-            }
-        } else {
-            try {
-                if (!o.isCancelled()) {
-                    subscriber.onError(cause);
+        try {
+            final Throwable cause = o.cause();
+            if (cause == null) {
+                try {
+                    subscriber.onComplete();
+                } finally {
+                    closeFuture.complete(null);
+                }
+            } else {
+                try {
+                    if (!o.isCancelled()) {
+                        subscriber.onError(cause);
+                    }
+                } finally {
+                    closeFuture.completeExceptionally(cause);
                 }
-            } finally {
-                closeFuture.completeExceptionally(cause);
             }
+        } finally {
+            setState(State.CLEANUP);
```

Comment:
Could you move`setState(State.CLEANUP)` where it originally was? Otherwise, `Subscription.cancel()` called by `Subscriber.onComplete()` can add an extra `CloseEvent` to the queue.